tracing-error/src/stack_error.rs

```diff
@@ -0,0 +1,192 @@
+                vtable,
+                spantrace: SpanTrace::capture(),
+                error,
+            },
+        }
+    }
+}
+
+#[repr(C)]
+struct ErrorImpl<E> {
+    vtable: &'static ErrorVTable,
+    spantrace: SpanTrace,
+    // NOTE: Don't use directly. Use only through vtable. Erased type may have
+    // different alignment.
+    error: E,
+}
+
+impl ErrorImpl<Erased> {
+    pub(crate) fn error(&self) -> &(dyn Error + Send + Sync + 'static) {
+        // # SAFETY
+        //
+        // this function is used to cast a type-erased pointer to a pointer to error's
+        // original type. the `ErrorImpl::error` method, which calls this function, requires that
+        // the type this function casts to be the original erased type of the error; failure to
+        // uphold this is UB. since the `From` impl is parameterized over the original error type,
+        // the function pointer we construct here will also retain the original type. therefore,
+        // when this is consumed by the `error` method, it will be safe to call.
+        unsafe { &*(self.vtable.object_ref)(self) }
+    }
+}
+
+struct ErrorVTable {
+    object_ref: unsafe fn(&ErrorImpl<Erased>) -> &(dyn Error + Send + Sync + 'static),
+}
+
+// # SAFETY
+//
+// This function must be parameterized on the type E of the original error that is being stored
+// inside of the `ErrorImpl`. When it is correctly parameterized it defines a function that safely
+// casts the Erased ErrorImpl pointer type back to the original pointer type.
+unsafe fn object_ref<E>(e: &ErrorImpl<Erased>) -> &(dyn Error + Send + Sync + 'static)
+where
+    E: Error + Send + Sync + 'static,
+{
+    // Attach E's native Error vtable onto a pointer to self.error.
+    &(*(e as *const ErrorImpl<Erased> as *const ErrorImpl<E>)).error
+}
+
+impl<E> Error for TracedError<E> {
+    // # SAFETY
+    //
+    // This function is safe so long as all functions on `ErrorImpl<Erased>` only ever access the
+    // wrapped error type via the `error` method defined on `ErrorImpl<Erased>`, which uses the
+    // function in the vtable to safely convert the pointer type back to the original type then
+    // returns the reference to the internal error.
+    //
+    // This function is necessary for the `downcast_ref` in `SpanTraceExtract` to work, because it
+    // needs a concrete type to downcast to and we cannot downcast to ErrorImpls parameterized on
+    // errors defined in other crates. By erasing the type here we can always cast back to the
+    // Erased version of the ErrorImpl pointer and still access the internal error type safely
+    // through the vtable.
+    fn source<'a>(&'a self) -> Option<&'a (dyn Error + 'static)> {
+        let erased = unsafe { &*(&self.inner as *const ErrorImpl<E> as *const ErrorImpl<Erased>) };
+        Some(erased)
+    }
+}
+
+impl<E> Debug for TracedError<E> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "TRACED ERROR PLACEHOLDER")
```

Comment:
we should probably replace this with a real `Debug` impl before merging?

Reply:
There really isn't a concept of a "real debug or display" impl. This is a hack that doesn't actually represent an error. We could leave it empty, or we could put a nicer message, but all in all this debug impl only exists so that it can implement error so that it can transmute the pointer to itself in the next call to `source`

Reply:
Ahh, I get it, right. That makes sense. Hmm, if this should never be called normally, perhaps it should be `unreachable!()`? If this gets called, that's a bug, right?

Reply:
no, it always has to be called, the way the stack error works is basically by returning itself twice in the error chain, but theres no way to communicate to the person iterating thru the error chain that they're on the fake error version of the stack error (the non type erased one) vs being on the real error one that has been type erased and can actually extract the spantrace.

ON SECOND THOUGHT, I should have the fake error be the second one, the one that returns the SpanTrace, because then the implementor can use the fact that it successfully returned a spantrace as an indicator that the error should not be rendered in the error chain and it can just skip straight to the next source.

Reply:
okay, so what i'm trying to suss out is: will this string ever be printed to a user?

Reply:
its unavoidable that it _could_ be, but if paired with an error wrapping type that expects TracedErrors it should be easy to skip. So... no?